update report (#512)
JuliaReinforcementLearning/ReinforcementLearning.jl: docs/homepage/blog/phase2_technical_report_of_enriching_offline_reinforcement_learning_algorithms_in_reinforcement_learning_jl/index.md

- Time planning: the following is a relatively simple time table.
  
| Date       | Work    |
| :-:       | :-:   |
| Prior - June 30   | Preliminary research, including algorithm papers, ReinforcementLearning.jl library code, etc. |
| The first phase |      |
| July1 - July15  | Design and build the framework of offline RL.  |
| July16 - July31 | Implement and experiment offline DQN and  offline SAC as benchmark. |
| August1 - August15 | Write build-in documentation and technical report. Implement and experiment CRR.|
| The second phase |      |
| August16 - August31 | Implement and experiment PLAS. |
| September1 - September15  | Research, implement and experiment new SOTA offline RL algorithms. |
| September16 - September30 | Write build-in documentation and technical report. Buffer for unexpected delay. |

> | After project | Carry on fixing issues and maintain implemented algorithms.   |

Reviewer: ```suggestion
| After project | Carry on fixing issues and maintaining implemented algorithms.   |
```

## Completed Work

It summarizes all the results of the second phase.
#### Offline RL algorithms

> In the second phase, we finish the offline RL algorithms: BCQ, Discrete-BCQ, BEAR(UWAC), FisherBRC. CRR and PLAS are detailed in the report of the first phase. Except for algorithms in discrete action space (such as Discrete-BCQ, Discrete-CRR), all algorithms can run on CPU and GPU, because `CartesianIndex` is not supported in GPU.

Reviewer: ```suggestion
In the second phase, we finished the offline RL algorithms: BCQ, Discrete-BCQ, BEAR(UWAC), FisherBRC. CRR and PLAS are detailed in the report of the first phase. All implemented algorithms except for those with discrete action space (such as Discrete-BCQ, Discrete-CRR) can run on both CPU and GPU. The main reason that some algorithms are not working on GPU is that,  `CartesianIndex` related operation is not well supported on GPU yet.
```

    ### Omit other parameters
    policy::BA1
    target_policy::BA2
    qnetwork1::BC1
    qnetwork2::BC2
    target_qnetwork1::BC1
    target_qnetwork2::BC2
    vae::V
    p::Int
    start_step::Int
end
```
In BCQ, we use `PerturbationNetwork`: $\pi(s,a)\rightarrow a$ to model policy and use $Q(s,a)$ to model Q-network.

`p` is a hyper-parameter used for repeating states to obtain better Q estimation. For example, BCQ repeats states and obtains the actions and Q-values. Then it chooses an action with the highest Q-value.
```julia
function (l::BCQLearner)(env)
    s = send_to_device(device(l.policy), state(env))
    s = Flux.unsqueeze(s, ndims(s) + 1)
    s = repeat(s, outer=(1, 1, l.p))
    action = l.policy(s, decode(l.vae.model, s))
    q_value = l.qnetwork1(vcat(s, action))
    idx = argmax(q_value)
    action[idx]
end
```
`start_step` represents the steps to train VAE alone.

Performance curve of BCQ algorithm in Pendulum (`start_step=100`):

\dfig{body;JuliaRL_BCQ_Pendulum.png}

##### Discrete BCQ
Discrete BCQ\dcite{fujimoto2019benchmarking} is a simple offline RL algorithm in discrete action space. Its pseudocode is as follow:

\dfig{body;DBCQ.png}

> The core of Discrete BCQ is calculating a mask of actions. We calculate the probability of actions in a given state, and divide by the maximum probability value. Then, we compare this value with 

Reviewer: ```suggestion
The core of Discrete BCQ is calculating a mask of actions. We calculate the probability of actions in a given state, and divide it by the maximum probability value. Then, we compare this value with 
```

```julia
mutable struct BCQDLearner{Aq, At} <: AbstractLearner
    ### Omit other parameters
    approximator::Aq
    target_approximator::At
    θ::Float32
    threshold::Float32
end
```

In Discrete BCQ, we use the Actor-Critic structure. The Critic is modeled as $Q(s,\cdot)$, and the Actor is modeled as $L(s,\cdot)$ (likelihood of the state).

`θ` is the regularization coefficient. When calculating the loss in Discrete BCQ, it uses likelihood of the state to regularize:
```julia
logit = AC.actor(s)
actor_loss + critic_loss + θ * mean(logit .^ 2)
```

Performance curve of Discrete BCQ algorithm in CartPole:

\dfig{body;JuliaRL_BCQD_CartPole.png}

##### BEAR & UWAC
Bootstrapping error accumulation reduction (BEAR) \dcite{DBLP:conf/nips/KumarFSTL19} is a policy-constraint method, which uses support constrain instead of distribution matching in BCQ. Its pseudocode is as follow:

\dfig{body;BEAR.png}
![avatar](BEAR.png)

The equation 1:

$$\pi_\phi := \max_{\pi \in \triangle_{|S|}}\mathbb{E}_{s\in \mathcal{D}}\mathbb{E}_{a\sim\pi(\cdot|s)}\big[\min_{j=1,\cdots,K}\hat{Q}_j(s,a)\big] \\ s.t. \mathbb{E}_{s\in\mathcal{D}}[{\rm MMD}(\mathcal{D}(s),\pi(\cdot|s))]\leq \varepsilon$$

> The Actor update is the main improvement part of BEAR. We needs to train a VAE to simulate sampling action from dataset. The training of the VAE and the learner are synchronized. Then we sample action by the Actor and calculate maximum mean discrepancy (MMD) loss:

Reviewer: ```suggestion
The Actor update is the main improvement part of BEAR. We need to train a VAE to simulate sampling action from dataset. The training of the VAE and the learner are synchronized. Then we sample action by the Actor and calculate maximum mean discrepancy (MMD) loss:
```

    s, a, r, t = dataset_ex.policy.trajectory.traces
    for i in 1:dataset_size
        push!(dataset, JuliaRLTransition(s[:, i], a[i], r[i], t[i], s[:, i+1]))
    end
    dataset
end
```

Specifically, we first run the built-in data collection experiment. For CartPole environment, we use BasicDQN to collect dataset. For Pendulum, SAC is used. Then we turn the trajectory into a vector of `JuliaRLTranstion`.

We can run the [offline RL experiments](https://juliareinforcementlearning.org/docs/experiments/#Offline) with the following commands (list all supported commands):
```julia
run(E`JuliaRL_BCQ_Pendulum(type)`)
run(E`JuliaRL_BCQD_CartPole(type)`)
run(E`JuliaRL_BEAR_Pendulum(type)`)
run(E`JuliaRL_CRR_Pendulum(type)`)
run(E`JuliaRL_CRR_CartPole(type)`)
run(E`JuliaRL_FisherBRC_Pendulum(type)`)
run(E`JuliaRL_PLAS_Pendulum(type)`)
```
`type` represents the method of collecting dataset. `type` can be `random`, `medium` or `expert`. `random` represents dataset generated by random agent, and `medium` uses an agent trained from scratch to convergence, and `expert` uses an agent that has converged. For example:

```julia
run(E`JuliaRL_BCQ_Pendulum(medium)`)
```

## Conclusion

> The implemented algorithms include the mainstream of the policy constraint method in offline reinforcement learning (including distribution matching, support constrain, implicit constraint, behavior cloning). And these algorithms include discrete and continuous action spaces. Besides, we test it in different datasets.

Reviewer: ```suggestion
The implemented algorithms in this project contain most of the policy constraint methods in offline reinforcement learning (including distribution matching, support constrain, implicit constraint, behavior cloning). And these algorithms cover both discrete and continuous action spaces. Besides, we have them tested with various datasets.
```
